src/components/molecules/DataWrapper.tsx

```diff
@@ -0,0 +1,22 @@
+import React, { ReactNode, useEffect, useState } from 'react';
+import { ApolloError, DocumentNode, useQuery } from '@apollo/client';
+import { TimeEntry } from '../atoms/ThresholdChart';
+import { formatMetData } from '../../utils/formatMetData';
+
+export type DataWrapperProps = {
+    query: DocumentNode;
+    children: (args: { data: TimeEntry[] | undefined; loading: boolean; error: ApolloError | undefined }) => ReactNode;
```

Comment:
This limits the DataWrapper to only work with the `ThresholdChart` (only accepting `data: TimeEntry[]`).

I believe this would be a fitting time to introduce ✨ _generics_ ✨ 

In essence, you wish to create to create a `DataWrapper` which takes
- a `query` which resolves to a type `T` (for use in `useQuery<T>(query)`)
- a mapping function mapping from type `T -> U`, where `U` is the input to the child function
_What does this mean?_ Albeit a slightly functional (and specifically Haskell) explanation, it essentially means you have to provide a function (in the props) which takes the **query data type** (`T`, e.g. `MetApiRawData`) and maps it to **some arbitrary output type** (`U`). In your example, you have provided this, only hard-coded. 

`formatMetData` takes the raw data type `T` as input, and produces some output type `U` which is further used by the children (in this example, `TimeEntry[]`). 

Thus we can say that `formatMetData` is a mapping from `T -> U`.

If you further eliminate most data from the child function, so you're left with `children: (data: U) => ReactNode`, this can be typed quite neatly, and you have provided the following;

- a **generic** (:sparkles: :mage:) component which retrieves an arbitrary GraphQL data source and provides it to a child visualisation. Given that you handle the error/loading states in this component, the child component (read: not the child function) can have a prop type like `props: { data: TimeEntry[] }` (note that it's not nullable) as you guarantee that it will be provided with actual data.

The usage would thus look like;

```tsx
<DataWrapper
  mapping={(entry) => ({ x: entry.x, y: entry.y })}
  query={I_LOVE_THE_MONKEY_HEAD_QUERY}
>
  {(data) => <ThresholdChart data={data} />}
</DataWrapper>
```

Reply:
Further information:

If you don't feel comfortable with generics:
- https://www.typescriptlang.org/docs/handbook/generics.html

- Google "react generic components"
- https://wanago.io/2020/03/09/functional-react-components-with-generic-props-in-typescript/

Reply:
Thanks for the good tips! I will definitely look more into generic types:)